midend/predication.cpp

```diff
@@ -15,13 +15,26 @@ limitations under the License.
 */
 
 #include "predication.h"
+#include "frontends/p4/cloner.h"
 
 namespace P4 {
 
+const IR::Expression* Predication::clone(const IR::Expression* expression) {
+    // We often need to clone expressions.  This is necessary because
+    // in the end we will generate different code for the different clones of
+    // an expression.  This is most obvious if one clone is on the LHS and one
+    // on the RHS of an assigment.
```

Comment:
You shouldn't actually need to clone expressions to do this -- but many of the later passes are broken when they run into dags rather than strict trees.  We should fix those passes, but that is a larger problem...

Reply:
Maybe what we should do is to run a pass that converts dags into graphs when it is necessary.
Unfortunately I no longer remember all places where this may be important...
In general this may be important when you maintain a hashmap from a Node* to some other data structure; sometimes the node context is important, so mapping just the Node* is not sufficient.